src/coord/src/coord.rs

```diff
@@ -289,21 +295,39 @@ where
         Ok(coord)
     }
 
-    pub fn now(&mut self) -> Timestamp {
-        let mut t = self
-            .start_time
-            .elapsed()
-            .as_millis()
-            .try_into()
-            .expect("system time did not fit in u64");
+    /// Assign a timestamp for a read.
+    fn get_read_ts(&mut self) -> Timestamp {
+        let ts = self.get_ts();
+        self.last_op_was_read = true;
+        self.last_read_ts = ts;
+        ts
+    }
+
+    /// Assign a timestamp for a write. Writes following reads must ensure that they are assigned a
+    /// strictly larger timestamp to ensure they are not visible to any real-time earlier reads.
+    fn get_write_ts(&mut self) -> Timestamp {
+        if self.last_op_was_read {
+            self.last_op_was_read = false;
+            loop {
+                let ts = self.get_ts();
+                if ts > self.last_read_ts {
+                    break ts;
+                }
+            }
+        } else {
+            self.get_ts()
+        }
+    }
```

Comment:
I haven't thought about this at all, but what do you think about just forcibly incrementing the read timestamp, rather than spinning while we do? For example, if someone performs a sequence of 10 updates to the catalog, we don't have to wait 10ms, we could just allocate those times and make sure that subsequent reads use them. We would only block if we did a read that wasn't ready yet, and it wouldn't block the coordinator just the terminal session.

Insane? Wrong?